File: Week_05/G20190343020102/32.最长有效括号.js
/*
 * @lc app=leetcode.cn id=32 lang=javascript
 *
 * [32] 最长有效括号
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function(s) {
  if(s.length == 0) return 0
  var temp = [{type: 0, index: 0}]
  max = 0
  num = 0
  for(i=0;i<s.length;i++){
      if(s[i] == '(') temp.push({type: '(', index: i+1})
      else if(s[i] == ')' && temp[temp.length-1].type == '(') temp.pop()
      else temp.push({type: ')', index: i+1})
  }
  temp.push({type: 0, index: s.length+1})
  for(j=0;j<temp.length-1;j++){
      num = temp[j+1].index - temp[j].index - 1
      if (num>max) max = num
  }
  return max
};
// @lc code=end

File: Week_05/G20190343020094/32.最长有效括号.py
#
# @lc app=leetcode.cn id=32 lang=python3
#
# [32] 最长有效括号
#

# @lc code=start
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        if(not s):
            return 0
        res=0
        n=len(s)
        dp=[0]*n
        for i in range(1,len(s)):
            if(s[i]==")"):
                if(s[i-1]=="("):
                    dp[i]=dp[i-2]+2
                if(s[i-1]==")" and i-dp[i-1]-1>=0 and s[i-dp[i-1]-1]=="("):
                    dp[i]=dp[i-1]+dp[i-dp[i-1]-2]+2
                res=max(res,dp[i])
        return res


# @lc code=end
if __name__ == "__main__":
    s = ')('
    res = Solution().longestValidParentheses(s)
    print(res)